tests/functional/test_configured_endpoint_url.py

```diff
@@ -0,0 +1,279 @@
+from pathlib import Path
+from unittest import mock
+
+import pytest
+
+import botocore.configprovider
+import botocore.utils
+from botocore.compat import urlsplit
+from tests import ClientHTTPStubber
+
+ENDPOINT_TESTDATA_FILE = (
+    Path(__file__).parent
+    / 'data'
+    / "configured-endpoint-urls"
+    / "test-cases"
+    / "profile-tests.json"
+)
+
+
+def dict_to_ini_section(ini_dict, section_header):
+    section_str = f'[{section_header}]\n'
+    for key, value in ini_dict.items():
+        if isinstance(value, dict):
+            section_str += f"{key} =\n"
+            for new_key, new_value in value.items():
+                section_str += f"  {new_key}={new_value}\n"
+        else:
+            section_str += f"{key}={value}\n"
+    return section_str + "\n"
+
+
+def create_cases():
+    with open(ENDPOINT_TESTDATA_FILE) as f:
+        test_suite = json.load(f)['testSuites'][0]
+
+    for test_case_data in test_suite['endpointUrlTests']:
+        yield pytest.param(
+            {
+                'service': test_case_data['service'],
+                'profile': test_case_data['profile'],
+                'expected_endpoint_url': test_case_data['output'][
+                    'endpointUrl'
+                ],
+                'client_args': test_suite['client_configs'].get(
+                    test_case_data['client_config'], {}
+                ),
+                'config_file_contents': get_config_file_contents(
+                    test_case_data['profile'], test_suite
+                ),
+                'environment': test_suite['environments'].get(
+                    test_case_data['environment'], {}
+                ),
+            },
+            id=test_case_data['name'],
+        )
+
+
+def get_config_file_contents(profile_name, test_suite):
+    profile = test_suite['profiles'][profile_name]
+
+    profile_str = dict_to_ini_section(
+        profile,
+        section_header=f"profile {profile_name}",
+    )
+
+    services_section_name = profile.get('services', None)
+
+    if services_section_name is None:
+        return profile_str
+
+    services_section = test_suite['services'][services_section_name]
+
+    service_section_str = dict_to_ini_section(
+        services_section,
+        section_header=f'services {services_section_name}',
+    )
+
+    return profile_str + service_section_str
+
+
+@pytest.fixture
+def client_creator(tmp_path):
+    tmp_config_file_path = tmp_path / 'config'
+    environ = {'AWS_CONFIG_FILE': str(tmp_config_file_path)}
+
+    def _do_create_client(
+        service,
+        profile,
+        client_args=None,
+        config_file_contents=None,
+        environment=None,
+    ):
+        environ.update(environment)
+        tmp_config_file = open(tmp_config_file_path, 'w')
+        tmp_config_file.write(config_file_contents)
+        tmp_config_file.flush()
+
+        session = botocore.session.Session(profile=profile)
+        session.set_config_variable('profile', profile)
```

Comment:
Do we need to call `set_config_variable` if it is provided on the `Session` object?